### PR TITLE
fix(datasets): warn on agent SFT rows with no supervised tokens

### PR DESCRIPTION
**Pull request by khazic**
# What does this PR do ?

When truncation drops every assistant turn (or last-turn masking leaves
nothing), `_format_example` produces an agent SFT sample whose `labels` are all
`ignore_index` (-100). The pre-existing guard only catches "no assistant message
present", not "assistant tokens truncated away".

A *single* such sample is harmless: the loss normalizes by the batch's
supervised-token count, so an all-masked example contributes nothing to the
gradient. A NaN only arises if a *whole* step is empty, which signals a
misconfigured `seq_length`/`truncation`. So rather than hard-failing the run on
one example, this detects the all-masked case and **warns** with the example id
(pointing at `seq_length`/`truncation`) while **keeping** the sample — it is
already effectively skipped from the gradient, and the map-style
dataset/collator cannot drop a row in place at this stage.

# Changelog

- After rendering (and after optional last-turn masking), emit a warning tagged
  with the example `id` when a sample has no supervised tokens (all labels
  masked), pointing at `seq_length`/`truncation` as the likely cause. The sample
  is kept, not dropped or raised on.
- Add unit tests for the all-masked (warns, kept) and partially-supervised
  (passes through) cases.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?

# Additional Information

- Rebased onto current `main` (after #2361 moved the renderer body into
  `_format_example_impl` and #2360 added the `seq_length`-inert warning); both
  warnings coexist in `agent_chat.py`.

### CODE REVIEW

File: nemo_automodel/components/datasets/llm/agent_chat.py [hunk 1]
```diff
@@ -301,6 +301,18 @@ def _format_example(
     )
     if train_on_last_turn_only:
         _mask_labels_to_last_turn(tokenized["labels"])
+
+    # Truncation (or over-aggressive last-turn masking) can leave a sample with no
+    # supervised tokens at all — every label is ``ignore_index`` (-100). Such a
+    # sample contributes no gradient and can drive a NaN loss for the batch, so
+    # fail loudly with the example id instead of emitting it silently.
```
Comment (HuiyingLi) [inline nemo_automodel/components/datasets/llm/agent_chat.py]: What do you think about skipping example(s) with warning v.s. hard fail the run?

**Review by HuiyingLi — DISMISSED** [review]
Thank you so much

File: nemo_automodel/components/datasets/llm/agent_chat.py (hunk 1 above)
Comment (khazic) [reply nemo_automodel/components/datasets/llm/agent_chat.py]: Thanks — good call, switched from hard-fail to a warning in 302a6d39.

Reasoning: a single all-masked example is actually harmless. The loss normalizes by the batch's supervised-token count (`loss = loss / num_label_tokens`, e.g. `linear_ce.py`), so an empty sample just contributes 0 — a NaN only arises if a *whole* step has zero supervised tokens, which signals a systematic `seq_length`/`truncation` misconfig rather than one bad row. So aborting the run on a single example was too aggressive.

On skip vs keep: I warn (with the example id) and keep the sample rather than truly dropping it. Detection happens inside the lazy `map_fn`, where the row can't be dropped in place — `LazyMappedDataset` is fixed-length and `default_collater` doesn't accept `None` (the existing `skip_invalid_samples` only skips at JSONL-read time). Since the kept sample contributes nothing to the gradient, warn-and-keep is the clean equivalent of skipping here, and the per-sample warning still surfaces the systematic case (a flood of warnings → check `seq_length`).

The test now asserts the warning fires (carrying the id) and that the sample passes through unchanged.

**Review by HuiyingLi — DISMISSED** [review]
LGTM thank you so much.

**Review by HuiyingLi — APPROVED** [review]
